RyleyGG/apluschiever
# Frontend OAuth2 Implementation

# Issue
Closes #23 

# Description
Created an `auth` folder within the fronted to contain authentication related utilities/services/guards/etc. 

Created an OAuth2 service to interact with the FASTAPI server.

Created an OAuth2 interceptor to automatically add authentication token to outgoing HTTP requests.

Created various Guards to prevent unauthenticated users from accessing content they shouldn't be able to. Also created guards to ensure users which are already logged in cannot reach the registration/sign in pages, and so unauthenticated users can't access the sign out page. These guards can be further expanded to ensure only users of certain 'roles' can access certain content.

Created new routes `sign_up`, `sign_in`, `sign_out`, and `landing`. 

Created new pages for each of the above new routes (as a component), and extracted the old hello world template code into the landing page component.

# Type Of Change

# Test Steps
Tested signing up, signing in, and signing out with dummy test data in the angular application and it works fine (this dummy data is currently pushed so others can replicate/modfiy it).

CompoDoc shows documentation for newly added code segments is extremely high (except for the dummy page  components I made for the sign in, out, etc).

Tested the OAuth2Service by writing extensive unit tests which ensure that tokens are set into local storage correctly and that only the necessary api calls are being made. 

Validated automatic refresh of auth tokens is working by  setting access token valid time to 1 minute and testing to see that the auth token TTL is updating. 

# Checklist

## Formatting
Please use markdown in your pull request message. A useful summary of commands can be found [here](https://guides.github.com/pdfs/markdown-cheatsheet-online.pdf).
